File: publisher.py
"""Local, unsigned DAT publication prototype. No upstream/network acquisition."""

import argparse
import copy
from datetime import datetime, timezone
from email.utils import format_datetime
import fcntl
import hashlib
import io
import json
import os
from pathlib import Path, PurePosixPath
import re
import tempfile
from urllib.parse import urlsplit
import xml.etree.ElementTree as XML
import zipfile

from defusedxml import ElementTree as SafeXML
from defusedxml.common import DefusedXmlException

MAX_INPUT = 16 * 1024 * 1024
MAX_DOCUMENT = 32 * 1024 * 1024
MAX_METADATA = 16 * 1024 * 1024
CATALOG_ID = re.compile(r"[a-z0-9]+(?:[./-][a-z0-9]+)*\Z")
OBJECT_PATH = re.compile(r"objects/[0-9a-f]{64}\.(?:dat|json|xml)\Z")
NS = "urn:romd:dat-catalog:experimental:1"
XML.register_namespace("romd", NS)


class InvalidCandidate(ValueError):
    pass


def encode(value):
    return (json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False) + "\n").encode()


def digest(data):
    return hashlib.sha256(data).hexdigest()


def bounded_read(path, limit):
    with Path(path).open("rb") as stream:
        data = stream.read(limit + 1)
    if len(data) > limit:
        raise InvalidCandidate("size_limit")
    return data


def document(raw, expected_name):
    """Preserve source bytes; allow only one regular DAT in a bounded ZIP."""
    if len(raw) > MAX_INPUT:
        raise InvalidCandidate("size_limit")
    try:
        if zipfile.is_zipfile(io.BytesIO(raw)):
            with zipfile.ZipFile(io.BytesIO(raw)) as archive:
                members = archive.infolist()
                if len(members) != 1:
                    raise InvalidCandidate("ambiguous_archive")
                member = members[0]
                path = PurePosixPath(member.filename)
                mode = member.external_attr >> 16
                if (member.is_dir() or path.is_absolute() or ".." in path.parts
                        or "\\" in member.filename or path.suffix.lower() != ".dat"
                        or (mode & 0o170000) == 0o120000 or member.flag_bits & 1):
                    raise InvalidCandidate("unsafe_archive")
                if member.file_size > MAX_DOCUMENT:
                    raise InvalidCandidate("size_limit")
                with archive.open(member) as stream:
                    raw = stream.read(MAX_DOCUMENT + 1)
        if len(raw) > MAX_DOCUMENT:
            raise InvalidCandidate("size_limit")
        root = SafeXML.fromstring(raw, forbid_entities=True, forbid_external=True)
        if root.tag != "datafile" or len(root.findall("header")) != 1:
            raise InvalidCandidate("invalid_dat")
        if root.findtext("header/name") != expected_name:
            raise InvalidCandidate("identity_mismatch")
        games = root.findall("game")
        names = [game.get("name") for game in games]
        if not names:
            raise InvalidCandidate("empty_dat")
        if any(not name for name in names) or len(set(names)) != len(names):
            raise InvalidCandidate("invalid_game_identity")
        return raw, {"games": len(games), "roms": len(root.findall("game/rom"))}
    except (DefusedXmlException, XML.ParseError, zipfile.BadZipFile, RuntimeError,
            NotImplementedError, EOFError) as error:
        raise InvalidCandidate("invalid_dat") from error


def atomic_write(path, data):
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temporary = tempfile.mkstemp(prefix=".upload-", dir=path.parent)
    try:
        with os.fdopen(fd, "wb") as stream:
            stream.write(data)
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def put_object(root, data, suffix):
    reference = {"path": f"objects/{digest(data)}.{suffix}", "sha256": digest(data), "bytes": len(data)}
    target = root / reference["path"]
    if target.exists():
        if target.read_bytes() != data:
            raise ValueError("immutable object corruption")
    else:
        atomic_write(target, data)
    return reference


def verified_read(root, reference):
    if not OBJECT_PATH.fullmatch(reference["path"]):
        raise ValueError("invalid object path")
    data = bounded_read(root / reference["path"], max(MAX_METADATA, MAX_DOCUMENT))
    if len(data) != reference["bytes"] or digest(data) != reference["sha256"]:
        raise ValueError("object integrity failure")
    return data


def load_snapshot(root):
    """Integrity checks only: this does NOT authenticate a remote publisher."""
    root = Path(root)
    current = json.loads(bounded_read(root / "current.json", MAX_METADATA))
    if current.get("format") != "romd-dat-catalog-experimental-1" or current.get("trust") != "unsigned-development":
        raise ValueError("unsupported metadata")
    snapshot = json.loads(verified_read(root, current["snapshot"]))
    if snapshot["sequence"] != current["sequence"]:
        raise ValueError("sequence mismatch")
    verified_read(root, snapshot["feed"])
    for catalog in snapshot["catalogs"].values():
        if catalog["artifact"]:
            verified_read(root, catalog["artifact"])
    for event in snapshot["events"]:
        verified_read(root, event["artifact"])
    return snapshot


def catch_up(snapshot, installed, rejected=None):
    """Reference fixture consumer: latest eligible version, independent of RSS."""
    rejected = rejected or {}
    return {key: value["artifact"] for key, value in snapshot["catalogs"].items()
            if value["artifact"] and installed.get(key) != value["artifact"]["sha256"]
            and value["artifact"]["sha256"] not in rejected.get(key, [])}


def feed_bytes(events, base_url):
    rss = XML.Element("rss", version="2.0")
    channel = XML.SubElement(rss, "channel")
    for tag, value in [("title", "ROMD DAT catalogs (development)"), ("link", base_url),
                       ("description", "Synthetic, unsigned DAT publication prototype")]:
        XML.SubElement(channel, tag).text = value
    for event in events:
        item = XML.SubElement(channel, "item")
        XML.SubElement(item, "title").text = event["name"]
        XML.SubElement(item, "guid", isPermaLink="false").text = event["id"]
        XML.SubElement(item, "pubDate").text = format_datetime(datetime.fromisoformat(event["publishedAt"]), usegmt=True)
        url = base_url + event["artifact"]["path"]
        XML.SubElement(item, "link").text = url
        XML.SubElement(item, "enclosure", url=url, length=str(event["artifact"]["bytes"]), type="application/xml")
        XML.SubElement(item, f"{{{NS}}}catalogId").text = event["catalogId"]
        XML.SubElement(item, f"{{{NS}}}sha256").text = event["artifact"]["sha256"]
        XML.SubElement(item, f"{{{NS}}}sequence").text = str(event["sequence"])
    return XML.tostring(rss, encoding="utf-8", xml_declaration=True)


def publish(root, attempts, base_url, *, now=None, feed_limit=100, before_commit=None):
    """Accept trusted local registry attempts; commit references only after objects exist."""
    root = Path(root)
    parsed = urlsplit(base_url)
    if parsed.scheme != "https" or not parsed.netloc or parsed.query or parsed.fragment or parsed.username or parsed.password:
        raise ValueError("base URL must be an HTTPS directory without credentials/query/fragment")
    base_url = base_url.rstrip("/") + "/"
    if not 1 <= feed_limit <= 1000:
        raise ValueError("feed_limit must be 1..1000")
    moment = now or datetime.now(timezone.utc)
    if moment.tzinfo is None:
        raise ValueError("timestamp must have a timezone")
    stamp = moment.astimezone(timezone.utc).isoformat()
    ids = [attempt["catalogId"] for attempt in attempts]
    if len(set(ids)) != len(ids) or any(not CATALOG_ID.fullmatch(key) for key in ids):
        raise ValueError("invalid or duplicate catalog ID")
    for attempt in attempts:
        if not attempt.get("expectedName") or not attempt.get("sourceUrl"):
            raise ValueError("registry requires expectedName and sourceUrl")
        source = urlsplit(attempt["sourceUrl"])
        if (source.scheme not in ("http", "https") or not source.netloc
                or source.username or source.password or source.query or source.fragment):
            raise ValueError("sourceUrl must be a public attribution URL without credentials/query/fragment")
        if ("path" in attempt) == ("failure" in attempt):
            raise ValueError("provide a local path or failure, exclusively")
    root.mkdir(parents=True, exist_ok=True)
    # Never unlink the lock: concurrent owners must keep the same inode.
    with (root / ".publish.lock").open("a") as lock:
        fcntl.flock(lock, fcntl.LOCK_EX)
        old = load_snapshot(root) if (root / "current.json").exists() else {"sequence": 0, "catalogs": {}, "events": []}
        if old.get("publishedAt", stamp) > stamp:
            raise ValueError("publication clock moved backwards")
        catalogs = copy.deepcopy(old["catalogs"])
        events = []
        sequence = old["sequence"] + 1
        for attempt in sorted(attempts, key=lambda x: x["catalogId"]):
            key = attempt["catalogId"]
            catalog = catalogs.get(key, {"name": attempt["expectedName"], "artifact": None,
                                       "lastSuccessfulCheck": None, "lastChanged": None})
            if catalog["name"] != attempt["expectedName"]:
                raise ValueError("registry identity changed; explicit migration required")
            catalog["lastAttempt"] = stamp
            try:
                if "failure" in attempt:
                    raise InvalidCandidate("acquisition_failed")
                data, counts = document(bounded_read(attempt["path"], MAX_INPUT), attempt["expectedName"])
                artifact = put_object(root, data, "dat")
                if catalog["artifact"] != artifact:
                    event = {"id": f"urn:romd:dat:{sequence}:{key}:{artifact['sha256']}",
                             "catalogId": key, "name": catalog["name"], "sequence": sequence,
                             "publishedAt": stamp, "artifact": artifact}
                    events.append(event)
                    catalog.update(artifact=artifact, lastChanged=stamp, counts=counts,
                                   provenance={"sourceUrl": attempt["sourceUrl"], "acquiredAt": stamp})
                catalog.update(lastSuccessfulCheck=stamp, health="healthy", error=None)
            except (InvalidCandidate, OSError) as error:
                # Do not publish local paths, credential-bearing URLs, or raw provider errors.
                catalog.update(health="failed", error=str(error) if isinstance(error, InvalidCandidate) else "io_failure")
            catalogs[key] = catalog
        events = (events + old["events"])[:feed_limit]
        feed = put_object(root, feed_bytes(events, base_url), "xml")
        snapshot = {"format": "romd-dat-catalog-experimental-1", "sequence": sequence,
                    "publishedAt": stamp, "catalogs": catalogs, "events": events, "feed": feed}
        reference = put_object(root, encode(snapshot), "json")
        # Validate every object referenced by the candidate before committing it.
        verified_read(root, reference)
        verified_read(root, feed)
        for value in catalogs.values():
            if value["artifact"]:
                verified_read(root, value["artifact"])
        for event in events:
            verified_read(root, event["artifact"])
        if before_commit:
            before_commit()
        atomic_write(root / "current.json", encode({"format": snapshot["format"],
                     "trust": "unsigned-development", "sequence": sequence, "snapshot": reference}))
        return snapshot


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("manifest", type=Path)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--base-url", default="https://catalogs.example.invalid/")
    args = parser.parse_args()
    attempts = json.loads(bounded_read(args.manifest, MAX_METADATA))
    for attempt in attempts:
        if "path" in attempt:
            attempt["path"] = args.manifest.parent / attempt["path"]
    result = publish(args.output, attempts, args.base_url)
    print(json.dumps({"sequence": result["sequence"], "catalogs": len(result["catalogs"]),
                      "failed": sum(c["health"] == "failed" for c in result["catalogs"].values()),
                      "trust": "unsigned-development"}))


if __name__ == "__main__":
    main()

File: internal/publisher/publisher.go
package publisher

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

const Format = "romd-dat-catalog-experimental-1"
const Trust = "unsigned-development"

var catalogID = regexp.MustCompile(`^[a-z0-9]+([./-][a-z0-9]+)*$`)
var objectPath = regexp.MustCompile(`^objects/[0-9a-f]{64}\.(dat|json|xml)$`)

type Reference struct {
	Path   string `json:"path"`
	SHA256 string `json:"sha256"`
	Bytes  int    `json:"bytes"`
}
type Attempt struct {
	RetryAt      *string `json:"retryAt,omitempty"`
	CatalogID    string  `json:"catalogId"`
	ExpectedName string  `json:"expectedName"`
	SourceURL    string  `json:"sourceUrl"`
	Path         *string `json:"path,omitempty"`
	Failure      *string `json:"failure,omitempty"`
}
type Provenance struct {
	SourceURL  string `json:"sourceUrl"`
	AcquiredAt string `json:"acquiredAt"`
}
type Catalog struct {
	RetryAt             *string     `json:"retryAt,omitempty"`
	Name                string      `json:"name"`
	Artifact            *Reference  `json:"artifact"`
	LastSuccessfulCheck *string     `json:"lastSuccessfulCheck"`
	LastChanged         *string     `json:"lastChanged"`
	LastAttempt         string      `json:"lastAttempt"`
	Health              string      `json:"health"`
	Error               *string     `json:"error"`
	Counts              *Counts     `json:"counts,omitempty"`
	Provenance          *Provenance `json:"provenance,omitempty"`
}
type Event struct {
	ID          string    `json:"id"`
	CatalogID   string    `json:"catalogId"`
	Name        string    `json:"name"`
	Sequence    int64     `json:"sequence"`
	PublishedAt string    `json:"publishedAt"`
	Artifact    Reference `json:"artifact"`
}
type Snapshot struct {
	Format      string             `json:"format"`
	Sequence    int64              `json:"sequence"`
	PublishedAt string             `json:"publishedAt"`
	Catalogs    map[string]Catalog `json:"catalogs"`
	Events      []Event            `json:"events"`
	Feed        Reference          `json:"feed"`
}
type Current struct {
	Format   string    `json:"format"`
	Trust    string    `json:"trust"`
	Sequence int64     `json:"sequence"`
	Snapshot Reference `json:"snapshot"`
}
type Options struct {
	Now          time.Time
	FeedLimit    int
	BeforeCommit func() error
}

func Hash(b []byte) string         { return fmt.Sprintf("%x", sha256.Sum256(b)) }
func encode(v any) ([]byte, error) { b, e := json.Marshal(v); return append(b, '\n'), e }
func atomicWrite(name string, b []byte) error {
	if e := os.MkdirAll(filepath.Dir(name), 0755); e != nil {
		return e
	}
	f, e := os.CreateTemp(filepath.Dir(name), ".upload-")
	if e != nil {
		return e
	}
	defer os.Remove(f.Name())
	defer f.Close()
	if _, e = f.Write(b); e != nil {
		return e
	}
	if e = f.Sync(); e != nil {
		return e
	}
	if e = f.Close(); e != nil {
		return e
	}
	return os.Rename(f.Name(), name)
}
func put(root string, b []byte, suffix string) (Reference, error) {
	r := Reference{"objects/" + Hash(b) + "." + suffix, Hash(b), len(b)}
	name := filepath.Join(root, r.Path)
	old, e := readFile(name, MaxDocument)
	if e == nil {
		if !bytes.Equal(old, b) {
			return r, errors.New("immutable object corruption")
		}
		return r, nil
	}
	if !errors.Is(e, os.ErrNotExist) {
		return r, e
	}
	return r, atomicWrite(name, b)
}
func VerifiedRead(root string, r Reference) ([]byte, error) {
	if !objectPath.MatchString(r.Path) {
		return nil, errors.New("invalid object path")
	}
	b, e := readFile(filepath.Join(root, r.Path), MaxDocument)
	if e != nil {
		return nil, e
	}
	if len(b) != r.Bytes || Hash(b) != r.SHA256 {
		return nil, errors.New("object integrity failure")
	}
	return b, nil
}
func verifySnapshot(root string, s Snapshot) error {
	if _, e := VerifiedRead(root, s.Feed); e != nil {
		return e
	}
	for _, c := range s.Catalogs {
		if c.Artifact != nil {
			if _, e := VerifiedRead(root, *c.Artifact); e != nil {
				return e
			}
		}
	}
	for _, ev := range s.Events {
		if _, e := VerifiedRead(root, ev.Artifact); e != nil {
			return e
		}
	}
	return nil
}

// LoadSnapshot checks integrity, not authenticity of a remote publisher.
func LoadSnapshot(root string) (Snapshot, error) {
	var c Current
	var s Snapshot
	b, e := readFile(filepath.Join(root, "current.json"), MaxMetadata)
	if e != nil {
		return s, e
	}
	if e = json.Unmarshal(b, &c); e != nil {
		return s, e
	}
	if c.Format != Format || c.Trust != Trust {
		return s, errors.New("unsupported metadata")
	}
	b, e = VerifiedRead(root, c.Snapshot)
	if e != nil {
		return s, e
	}
	if len(b) > MaxMetadata {
		return s, errors.New("metadata size limit")
	}
	if e = json.Unmarshal(b, &s); e != nil {
		return s, e
	}
	if s.Format != Format || s.Sequence != c.Sequence || s.Sequence < 1 || s.Catalogs == nil {
		return s, errors.New("invalid snapshot")
	}
	return s, verifySnapshot(root, s)
}
func CatchUp(s Snapshot, installed map[string]string, rejected map[string][]string) map[string]Reference {
	result := map[string]Reference{}
	for key, c := range s.Catalogs {
		if c.Artifact == nil || installed[key] == c.Artifact.SHA256 {
			continue
		}
		skip := false
		for _, hash := range rejected[key] {
			if hash == c.Artifact.SHA256 {
				skip = true
			}
		}
		if !skip {
			result[key] = *c.Artifact
		}
	}
	return result
}
func publicURL(value string, httpsOnly bool) bool {
	u, e := url.Parse(value)
	return e == nil && u.Hostname() != "" && u.User == nil && u.RawQuery == "" && !u.ForceQuery && u.Fragment == "" && (u.Scheme == "https" || (!httpsOnly && u.Scheme == "http"))
}

// Permit only the public Datomatic selection query, never session/download
// tokens or arbitrary query parameters that could carry credentials.
var datomaticSource = regexp.MustCompile(`^https://datomatic\.no-intro\.org/index\.php\?page=download&op=dat&s=[1-9][0-9]{0,5}$`)

func publicSourceURL(value string) bool {
	return publicURL(value, false) || datomaticSource.MatchString(value)
}

func Publish(root string, attempts []Attempt, baseURL string, opt Options) (Snapshot, error) {
	var empty Snapshot
	if !publicURL(baseURL, true) {
		return empty, errors.New("base URL must be public HTTPS without credentials/query/fragment")
	}
	baseURL = strings.TrimRight(baseURL, "/") + "/"
	if opt.FeedLimit == 0 {
		opt.FeedLimit = 100
	}
	if opt.FeedLimit < 1 || opt.FeedLimit > 1000 {
		return empty, errors.New("feed limit must be 1..1000")
	}
	if opt.Now.IsZero() {
		opt.Now = time.Now()
	}
	stamp := opt.Now.UTC().Format(time.RFC3339Nano)
	seen := map[string]bool{}
	for _, a := range attempts {
		if !catalogID.MatchString(a.CatalogID) || seen[a.CatalogID] {
			return empty, errors.New("invalid or duplicate catalog ID")
		}
		if a.RetryAt != nil {
			if _, err := time.Parse(time.RFC3339Nano, *a.RetryAt); err != nil || a.Failure == nil {
				return empty, errors.New("invalid acquisition retry time")
			}
		}
		seen[a.CatalogID] = true
		if a.ExpectedName == "" || !publicSourceURL(a.SourceURL) || (a.Path == nil) == (a.Failure == nil) {
			return empty, errors.New("invalid registry attempt")
		}
	}
	if e := os.MkdirAll(root, 0755); e != nil {
		return empty, e
	}
	unlock, e := lockPublisher(root)
	if e != nil {
		return empty, e
	}
	defer unlock()
	old, e := LoadSnapshot(root)
	if e != nil {
		if _, stat := os.Stat(filepath.Join(root, "current.json")); !errors.Is(stat, os.ErrNotExist) {
			return empty, e
		}
		old = Snapshot{Catalogs: map[string]Catalog{}, Events: []Event{}}
	}
	if old.PublishedAt != "" {
		previous, e := time.Parse(time.RFC3339Nano, old.PublishedAt)
		if e != nil || opt.Now.Before(previous) {
			return empty, errors.New("invalid or regressed publication clock")
		}
	}
	s := Snapshot{Format: Format, Sequence: old.Sequence + 1, PublishedAt: stamp, Catalogs: old.Catalogs, Events: []Event{}}
	ordered := append([]Attempt(nil), attempts...)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].CatalogID < ordered[j].CatalogID })
	for _, a := range ordered {
		c, exists := s.Catalogs[a.CatalogID]
		if !exists {
			c.Name = a.ExpectedName
		}
		if c.Name != a.ExpectedName {
			return empty, errors.New("registry identity changed; explicit migration required")
		}
		c.LastAttempt = stamp
		var raw []byte
		var counts Counts
		var candidateErr error
		if a.Failure != nil {
			candidateErr = CandidateError("acquisition_failed")
			if *a.Failure == "publication_paused" {
				candidateErr = CandidateError("publication_paused")
			}
		} else {
			raw, candidateErr = readFile(*a.Path, MaxInput)
			if candidateErr == nil {
				raw, counts, candidateErr = Document(raw, a.ExpectedName)
			}
		}
		if candidateErr != nil {
			code := "io_failure"
			var ce CandidateError
			if errors.As(candidateErr, &ce) {
				code = string(ce)
			}
			if a.RetryAt != nil {
				c.RetryAt = a.RetryAt
			}
			c.Health = "failed"
			c.Error = &code
		} else {
			artifact, e := put(root, raw, "dat")
			if e != nil {
				return empty, e
			}
			if c.Artifact == nil || *c.Artifact != artifact {
				s.Events = append(s.Events, Event{fmt.Sprintf("urn:romd:dat:%d:%s:%s", s.Sequence, a.CatalogID, artifact.SHA256), a.CatalogID, c.Name, s.Sequence, stamp, artifact})
				c.Artifact = &artifact
				c.LastChanged = &stamp
				c.Counts = &counts
				c.Provenance = &Provenance{a.SourceURL, stamp}
			}
			c.LastSuccessfulCheck = &stamp
			c.RetryAt = nil
			c.Health = "healthy"
			c.Error = nil
		}
		s.Catalogs[a.CatalogID] = c
	}
	s.Events = append(s.Events, old.Events...)
	if len(s.Events) > opt.FeedLimit {
		s.Events = s.Events[:opt.FeedLimit]
	}
	feed, e := feedBytes(s.Events, baseURL)
	if e != nil {
		return empty, e
	}
	s.Feed, e = put(root, feed, "xml")
	if e != nil {
		return empty, e
	}
	b, e := encode(s)
	if e != nil {
		return empty, e
	}
	if len(b) > MaxMetadata {
		return empty, errors.New("metadata size limit")
	}
	ref, e := put(root, b, "json")
	if e != nil {
		return empty, e
	}
	if _, e = VerifiedRead(root, ref); e != nil {
		return empty, e
	}
	if e = verifySnapshot(root, s); e != nil {
		return empty, e
	}
	if opt.BeforeCommit != nil {
		if e = opt.BeforeCommit(); e != nil {
			return empty, e
		}
	}
	b, e = encode(Current{Format, Trust, s.Sequence, ref})
	if e != nil {
		return empty, e
	}
	if e = atomicWrite(filepath.Join(root, "current.json"), b); e != nil {
		return empty, e
	}
	return s, nil
}

func ReadManifest(name string) ([]Attempt, error) {
	b, e := readFile(name, MaxMetadata)
	if e != nil {
		return nil, e
	}
	var a []Attempt
	if e = json.Unmarshal(b, &a); e != nil {
		return nil, e
	}
	if a == nil {
		return nil, errors.New("manifest must be an array")
	}
	for i := range a {
		if a[i].Path != nil && !filepath.IsAbs(*a[i].Path) {
			p := filepath.Join(filepath.Dir(name), *a[i].Path)
			a[i].Path = &p
		}
	}
	return a, nil
}
